utils/treesitter/grammar.js

```diff
@@ -0,0 +1,508 @@
+/*
+ * Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+ * Exceptions. See /LICENSE for license information.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+// This grammar is more permissive than toolchain because it is geared towards
+// editor use.
+
+function repeat_sep1(thing, sep) {
+  return seq(thing, repeat(seq(sep, thing)));
+}
+
+function comma_sep(thing) {
+  // Trailing comma is only allowed if there is at least one element.
+  return optional(seq(repeat_sep1(thing, ','), optional(',')));
+}
+
+// This is based on toolchain/parser/precedence.cpp
+const PREC = {
+  TermPrefix: 11,
+  TermPostfix: 11,
+  NumericPrefix: 10,
+  NumericPostfix: 10,
+  Multiplicative: 9,
+  Additive: 8,
+  BitwisePrefix: 7,
+  BitwiseAnd: 6,
+  BitwiseOr: 6,
+  BitwiseXor: 6,
+  BitShift: 6,
+  TypePostfix: 5,
+  LogicalPrefix: 4,
+  Relational: 3,
+  LogicalAnd: 2,
+  LogicalOr: 2,
+  WhereClause: 1,
+  IfExpression: 1,
+};
+
+module.exports = grammar({
+  name: 'carbon',
+
+  word: ($) => $.ident,
+
+  conflicts: ($) => [
+    [$.paren_pattern, $.paren_expression],
+    [$.struct_literal, $.struct_type_literal],
+  ],
+
+  extras: ($) => [/\s/, $.comment],
+
+  // NOTE: This must match the order in src/scanner.c, names are not used for matching.
+  externals: ($) => [$.binary_star, $.postfix_star],
+
+  rules: {
+    source_file: ($) =>
+      seq(
+        optional($.package_directive),
+        repeat($.import_directive),
+        repeat($.declaration)
+      ),
+
+    api_or_impl: ($) => choice('api', 'impl'),
+
+    library_path: ($) => seq('library', $.string_literal),
+
+    package_directive: ($) =>
+      seq('package', $.ident, optional($.library_path), $.api_or_impl, ';'),
+
+    import_directive: ($) =>
+      seq('import', $.ident, optional($.library_path), ';'),
+
+    comment: ($) => token(seq('//', /.*/)),
+
+    ident: ($) => /[A-Za-z_][A-Za-z0-9_]*/,
+
+    bool_literal: ($) => choice('true', 'false'),
+
+    numeric_literal: ($) => {
+      // This is using variables because rules are not allowed in token.immediate and token.
```

Comment:
Still seem to have long lines in this file.

Reply:
I think javascript formatter has greater column limit by default. I have manually wrapped these for now.